### ISSUE TITLE
Backend container fails to build on Apple Silicon Mac

### ISSUE THREAD

**Issue by c-ehrlich** [freeCodeCamp/chapter#1045]
### Describe the bug:
`docker-compose build` currently fails on Apple Silicon Macs. I haven't been able to fully figure out _why_ it breaks from the error logs, but a bit of research suggests that there are known compatibility issues when using Prisma in certain Docker images on Apple Silicon Macs. 

I was able to get the project running by using the following `server.Dockerfile` (using the `node:16` image and installing `netcat`)

```dockerfile
FROM node:16

WORKDIR /usr/chapter/

# Bundle app source
COPY ./server ./server
COPY package*.json ./

# Install app dependencies
RUN apt-get update && apt-get -y dist-upgrade && apt install -y netcat 
RUN npm ci -w=server --ignore-scripts

EXPOSE 5000

WORKDIR /usr/chapter/server

CMD [ "npm","run","dev" ]
```

Just to be clear I'm not suggesting that this Dockerfile should be used, as it creates an unnecessarily large image - it's just the first thing that I got to work. But I do think that making the project work out of the box on all common platforms should be a goal.

### Tell us about your browser and operating system:
M1 MacBook Air, MacOS 12.2.1

**Comment by allella** [comment]
Alright, so is the `FROM node:16-alpine` the main problem, or is it the `apt install -y netcat` because those seem to be the main differences.

I don't have Mac-anything, but it's curious if the differences above work on older Macs too.

**Comment by c-ehrlich** [comment]
tldr: It's possible to add one line to `docker-compose.yml` to make it emulate x64 on other architectures such as Apple Silicon. This seems like a good fix until there is a cross platform alpine image.

---

Here are my results of researching this further:

* As far as I can tell, no amount of adding packages to `node:16-alpine` will work. I get stuck on the same step as:
https://github.com/prisma/prisma/issues/8478#issuecomment-1091713235

* Using the regular `node:16` image requires additional packages because it's a fundamentally different image that includes different stuff by default. It's also a very large image compared to alpine. It works, but I would suggest to not use it.

* There's the possibility of compiling from scratch. which means the alpine image will work on all platforms, but this slows down the build process significantly (15+ minutes on my machine).
https://github.com/prisma/prisma/issues/8478#issuecomment-1069783120
https://github.com/prisma/prisma/issues/8478#issuecomment-1068983390
I included it here for the sake of completeness, but I don't think it's a good idea.

* It's possible to modify `docker-compose.yml` to tell other architectures to emulate amd64. This means the current Dockerfile with the alpine image works on Apple Silicon. It's a bit slower, but M1 is fast enough that even with emulation overhead it's still fine for development purposes. 
**I believe this is the best compromise until there is a cross platform alpine image.**
```
# docker-compose.yml
# added 'platform'

  app:
    env_file: .env
    depends_on:
      - db
    platform: "linux/amd64"
    build: 
      context: .
      dockerfile: server.Dockerfile
    working_dir: /usr/chapter/server
    environment:
      - DATABASE_URL=postgres://${DB_USER}:${DB_PASSWORD}@db:5432/${DB_NAME}
      - EMAIL_HOST=mailhog
    ports:
      - "5000:5000"
```

* The final possibility is to change nothing in the code, but add instructions to the docs for Apple Silicon users on how to modify the Dockerfile to get the server to run. They would also need to make sure to keep those changes out of version control.

* I don't have an Intel Mac to test on, but it should behave the same as any other x64 platform as the issue is related to CPU Architecture, not host OS. Likewise, it seems Linux on ARM (for example AWS Graviton) has the same issues as Apple Silicon.

**Comment by allella** [comment]
Thanks for the detailed report.

@ojeytonwilliams @nhcarrigan @tomiiide @NorthDecoder @tomnoland have been involved with the Docker work and may have thoughts or be able to test how the proposed change to add `platform: "linux/amd64"` to the docker-compose.yml

**Comment by ojeytonwilliams** [comment]
@c-ehrlich at the moment, we mainly want the Dockerfiles to provide a decent developer experience. We can always maintain separate production images that are optimised. 

With that in mind, is `node:16` more appealing (since image size shouldn't be too important in development) or is emulation the way to go regardless?

I'm fine with trying either approach and agree entirely that we want it to work smoothly on as many platforms as we can readily support.

**Comment by c-ehrlich** [comment]
To be honest I'm not sure what is better. 

It's about performance either way - one choice is to have slightly reduced performance on Apple Silicon Macs due to emulation, the other is to have a significantly larger image size for everyone (38mb vs 330mb compressed for the base image according to DockerHub), which might also be a performance consideration for people developing on low end PCs or slow network connections.

I'll try to work on some other issue in Chapter next week, run the test suite, etc. to see if there is any problem (compatibility or perceived performance) with emulation. If not, I would suggest emulation for the time being. That being said if someone has a stronger opinion on this or wants Chapter to build on Apple Silicon Macs ASAP, I'm also fine with either approach.

**Comment by allella** [comment]
@c-ehrlich is the emulation only going to slow down the Mac Silicon since most other processors will already be amd64? Or, does adding that emulation line also cause a typical x64 CPU to operate in a different mode?

If it's only slowing down the new Macs, then it would seem that's better than making everyone else slow by downloading 10x the  images.

**Comment by c-ehrlich** [comment]
@allella it should not make a difference for CPUs that are already amd64 (Intel CPUs are also amd64) - what the `platform` command actually does is tell Docker to download the image for the specified platform/architecture, so amd64 would use this image anyway. Emulation is how Apple Silicon (and Windows on ARM) deals with non-native images.

I worked on a different part of Chapter last night. Anecdotally the server feels indistinguishable from native, which makes sense as it doesn't need to handle a lot of requests during dev/testing.

By the way Windows on ARM also has x64 emulation. Some of the Snapdragon SoCs are a bit slower than the M1, but they should still be able to emulate fine.

The only devices I'm not sure about is ARM-based Chromebooks or other ARM-based Linux machines. However the current containers already wouldn't build on those so at least fixing it on Apple Silicon and Windows on ARM without impacting other platforms feels like a good improvement for the time being.
